Fix for Specifying unified_job_template as ID doesn't work

ppavlovrus opened the pull request:
##### SUMMARY

related #14458

Summary: set unified_job_template as value for key id in search_fields if 
unified_job_template is digit


##### ISSUE TYPE
 - Bug

##### COMPONENT NAME
 - Collection

##### AWX VERSION
```
awx: 0.1.dev34000+g36b2ecb
```


##### ADDITIONAL INFORMATION

This issue described at https://github.com/ansible/awx/issues/14458

Commit:
Fix for awx.awx.schedule - Specifying unified_job_template as ID doesn't work #14458
Summary: set unified_job_template as value for key id in search_fields if unified_job_template is digit

Signed-off-by: Павел Павлов <ppavlov@astralinux.ru>

fosterseth commented on awx_collection/plugins/modules/schedule.py:
```diff
@@ -262,7 +262,10 @@ def main():
     unified_job_template_id = None
     if unified_job_template:
-        search_fields['name'] = unified_job_template
+        if unified_job_template.isdigit():
```
`module.get_one` has a parameter called `name_or_id`. Might be better to use that instead

here is an example from a different module

https://github.com/ansible/awx/blob/63e82aa4a3533eb431d08a7a0bb080d4201597aa/awx_collection/plugins/modules/host.py#L105

ppavlovrus replied:
Hello, thank you for review! I changed code and it use name_or_id now

Commit:
Fix for awx.awx.schedule - Specifying unified_job_template as ID doesn't work #14458
Summary: fix after review, using name_or_id=id parameter

Signed-off-by: Павел Павлов <ppavlov@astralinux.ru>

ppavlovrus committed:
Merge branch 'devel' into fix-awx-14458

fosterseth commented on awx_collection/plugins/modules/schedule.py:
```diff
@@ -262,8 +262,7 @@ def main():
     if unified_job_template:
-        search_fields['name'] = unified_job_template
-        unified_job_template_id = module.get_one('unified_job_templates', **{'data': search_fields})['id']
+        unified_job_template_id = module.get_one('unified_job_templates', name_or_id=id, **{'data': search_fields})['id']
```
name_or_id=id, -- `id` isn't a defined variable here

instead I think you want `name_or_id=unified_job_template`

Also, since `search_fields` is no longer defined, we don't need to include `**{'data': search_fields}`

ppavlovrus replied:
Maybe you may give me some advise about how get_one should work.
i try following fix
    search_fields = {}
    sched_search_fields = {}
    if organization:
        search_fields['organization'] = module.resolve_name_to_id('organizations', organization)
    unified_job_template_id = None
    if unified_job_template:
        unified_job_template_id = module.get_one('unified_job_templates', name_or_id=unified_job_template)['id']
        sched_search_fields['unified_job_template'] = unified_job_template_id

but after i try to execute playbook_fail_modified.yaml playbook, mentioned in comment below, i receive error

[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [Create a scheduled job in AWX 2] ***********************************************************************************************************************************************************************************************

TASK [Test schedule job w/ ID] *******************************************************************************************************************************************************************************************************
[DEPRECATION WARNING]: The lookup plugin 'awx.awx.schedule_rrule' was expected to return a list, got '<class 'str'>' instead. The lookup plugin 'awx.awx.schedule_rrule' needs to be changed to return a list. This will be an error
in Ansible 2.18. This feature will be removed in version 2.18. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'NoneType' object is not subscriptable
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/Users/pavelpavlov/.ansible/tmp/ansible-tmp-1721122986.4784498-5157-129940876681239/AnsiballZ_schedule.py\", line 107, in <module>\n    _ansiballz_main()\n  File \"/Users/pavelpavlov/.ansible/tmp/ansible-tmp-1721122986.4784498-5157-129940876681239/AnsiballZ_schedule.py\", line 99, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/Users/pavelpavlov/.ansible/tmp/ansible-tmp-1721122986.4784498-5157-129940876681239/AnsiballZ_schedule.py\", line 47, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.awx.awx.plugins.modules.schedule', init_globals=dict(_module_fqn='ansible_collections.awx.awx.plugins.modules.schedule', _modlib_path=modlib_path),\n  File \"<frozen runpy>\", line 226, in run_module\n  File \"<frozen runpy>\", line 98, in _run_module_code\n  File \"<frozen runpy>\", line 88, in _run_code\n  File \"/var/folders/fj/x8_z0lgd55g3js285tnczxxh0000gn/T/ansible_awx.awx.schedule_payload_tfdomdw6/ansible_awx.awx.schedule_payload.zip/ansible_collections/awx/awx/plugins/modules/schedule.py\", line 370, in <module>\n  File \"/var/folders/fj/x8_z0lgd55g3js285tnczxxh0000gn/T/ansible_awx.awx.schedule_payload_tfdomdw6/ansible_awx.awx.schedule_payload.zip/ansible_collections/awx/awx/plugins/modules/schedule.py\", line 265, in main\nTypeError: 'NoneType' object is not subscriptable\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

PLAY RECAP ***************************************************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

So, line 265 is unified_job_template_id = module.get_one('unified_job_templates', name_or_id=unified_job_template)['id']. This implies, as far as I understand, that get_one returns None. Of course, I am debugging this and will find out what happens, but is there an obvious error here?

fosterseth commented:
also can you provide steps on how you are verifying the change?

ppavlovrus committed:
Merge branch 'ansible:devel' into fix-awx-14458

Commit:
Fix for awx.awx.schedule - Specifying unified_job_template as ID doesn't work #14458
Summary: changes after review name_or_id=id changed to name_or_id=unified_jod_template

Signed-off-by: Павел Павлов <ppavlov@astralinux.ru>

ppavlovrus committed:
Merge branch 'ansible:devel' into fix-awx-14458

Commit:
Fix for awx.awx.schedule - Specifying unified_job_template as ID doesn't work #14458
Summary: remove  **{'data': search_fields}

Signed-off-by: Павел Павлов <ppavlov@astralinux.ru>

ppavlovrus commented:
> also can you provide steps on how you are verifying the change?

Finally manage to get back to this issue. I verify it by the following steps:
1) Using my branch build dev environment with command make docker-compose-build
2) Run it with make docker-compose
3) Start dev ui with command docker exec -ti tools_awx_1 make ui-devel-test
4) Create superuser, for example admin2 qith password awx - docker exec -ti tools_awx_1 awx-manage createsuperuser
5) Prepare environment variables for ansible-playbook with commands 
export CONTROLLER_HOST=https://127.0.0.1:3000
export CONTROLLER_USERNAME=admin2
export CONTROLLER_PASSWORD=awx
6) Build and install collection with make build_collection  make install_collection
7) After this run playbooks for success and fail result with following commands:
ansible-playbook  playbook_success_modified.yaml
ansible-playbook  playbook_fail_modified.yaml

 playbook_success_modified.yaml:
`
- name: Create a scheduled job in AWX 
  hosts: localhost
  gather_facts: no
  tasks:
    - name: Test schedule job w/ ID
      vars:
        start_time: "2025-05-02 19:10:00"
      awx.awx.schedule:
        name: "Test with name lalala"
        unified_job_template: 'Demo Job Template'
        validate_certs: false
        rrule: "{{ query('awx.awx.schedule_rrule', 'minute', start_date=start_time, timezone='UTC', every='1', end_on='1') }}"
`
 

playbook_fail_modified.yaml:
`
- name: Create a scheduled job in AWX
  hosts: localhost
  gather_facts: no
  tasks:
    - name: Test schedule job w/ ID
      vars:
        start_time: "2026-05-02 19:05:14"
      awx.awx.schedule:
        name: "Test with ID"
        unified_job_template: '17'
        validate_certs: false
        rrule: "{{ query('awx.awx.schedule_rrule', 'minute', start_date=start_time, timezone='UTC', every='1', end_on='1') }}"
`

ppavlovrus committed:
Merge branch 'ansible:devel' into fix-awx-14458

ppavlovrus commented:
will create another PR
